src/TgSharp.Core/Network/WSTransport.cs

```diff
@@ -0,0 +1,178 @@
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Net;
+using System.Net.WebSockets;
+using System.Reactive.Concurrency;
+using System.Reactive.Linq;
+using System.Security.Cryptography;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using TgSharp.Core.MTProto.Crypto;
+using Websocket.Client;
+using Websocket.Client.Models;
+
+namespace TgSharp.Core.Network
+{
+    public class WSTransport
+    {
+        private WebsocketClient client;
+        private const uint protocol = 0xeeeeeeee;
+        private readonly RandomNumberGenerator rngSource;
+        private readonly string address;
+        private readonly int port;
+        private byte[] encryptKey, encryptIV, decryptKey, decryptIV;
+
+        private byte[] encryptCount = new byte[16], decryptCount = new byte[16];
+        private int encryptNum = 0, decryptNum = 0;
+
+        public Action<Message> OnEncryptedMessage, OnUnencryptedMessage;
+        private static readonly object syncGate = new object();
+
+
+        public WSTransport(string address, int port)
+        {
+            Array.Clear(encryptCount, 0, encryptCount.Length);
+            Array.Clear(decryptCount, 0, decryptCount.Length);
+
+            rngSource = RandomNumberGenerator.Create();
+            this.address = address;
+            this.port = port;
+        }
+
+        private async Task InitializeClient()
+        {
+            client = new WebsocketClient(new Uri($"ws://{address}:{port}/apiws"));
```

Comment:
can you rather assign this in the constructor?

Reply:
no, there is a bug in the library we use (tbh I don't know if it's a bug or expected behavior) that when you call .StartOrFail() it sets a flag as true and throws in case of an error but when you want to retry it because that flag is already true it doesn't even try, now that I think about it I might need to dispose it before replacing it

Reply:
added a dispose call before it

Reply:
I only said assign, I didnt mean you to call .StartOrFail() in the constructor

Reply:
I call initializeClient in the retry loop for connecting, If I assign the client in constructor the bug that I just said will happen

Reply:
so I have to reassign it to a new one every time

Reply:
which should be inside the loop and that's the best place for it